src/librustc_mir/transform/simplify_cfg.rs

```diff
@@ -65,25 +44,70 @@ impl SimplifyCfg {
             Some(target)
         }
 
+        let mut predecessor_map = util::build_predecessor_map(mir);
+
         let mut changed = false;
-        for bb in mir.all_basic_blocks() {
+        let mut seen = vec![false; mir.basic_blocks.len()];
+        let mut worklist = vec![START_BLOCK];
+        while let Some(bb) = worklist.pop() {
             // Temporarily swap out the terminator we're modifying to keep borrowck happy
             let mut terminator = Terminator::Diverge;
             mem::swap(&mut terminator, &mut mir.basic_block_data_mut(bb).terminator);
 
+            // Shortcut chains of empty blocks that just jump from one to the next
             for target in terminator.successors_mut() {
                 let new_target = match final_target(mir, *target) {
                     Some(new_target) => new_target,
                     None if mir.basic_block_data(bb).statements.is_empty() => bb,
                     None => continue
                 };
-                changed |= *target != new_target;
-                *target = new_target;
+
+                if *target != new_target {
+                    changed = true;
+                    predecessor_map.remove_predecessor(*target);
+                    predecessor_map.add_predecessor(new_target);
+                    *target = new_target;
+                }
             }
 
-            mir.basic_block_data_mut(bb).terminator = terminator;
+            // See if we can merge the target block into this one
+            match terminator {
+                Terminator::Goto { target } if target.index() > DIVERGE_BLOCK.index() &&
+                        predecessor_map.num_predecessors(target) == 1 => {
+                    changed = true;
+                    let mut other_data = BasicBlockData {
+                        statements: Vec::new(),
+                        terminator: Terminator::Goto { target: target}
+                    };
+                    mem::swap(&mut other_data, mir.basic_block_data_mut(target));
+
+                    // target used to have 1 predecessor (bb), and still has only one (itself)
+                    // All the successors of target have had target replaced by bb in their
+                    // list of predecessors, keeping the number the same.
+
+                    let data = mir.basic_block_data_mut(bb);
+                    data.statements.append(&mut other_data.statements);
+                    mem::swap(&mut data.terminator, &mut other_data.terminator);
+                }
+                _ => mir.basic_block_data_mut(bb).terminator = terminator
```

Comment:
Nit: might be a little clearer if we move the assignment here after the match, so that it's clear that we restore the `terminator` on both paths (this took me a bit of puzzling out). But I'm not sure, and I guess it's a bit less efficient that way. Maybe just add a comment like `// restore the terminator we swapped out` on both branches?